app/models/submission.rb

```diff
@@ -32,5 +34,14 @@ def reject
   def average_rate
     rates.count == 0 ? 0 : (rates.sum(:value).to_f / rates.count)
   end
-end
 
+  def generate_confirmation_token!
+    self.confirmation_token = Devise.friendly_token
+    self.confirmation_token_created_at = Time.now
+    save
+  end
+
+  def has_expired?
+    awaiting? && confirmation_token_created_at < 1.week.ago
+  end
```

Comment:
I think this method with that name is misleading. If some submission is already marked as expired, then `submission.has_expired?` will return `false`.

You could change the name to `to_be_marked_as_expired`. Personally, I would move the logic into the service which is using that method.